python/test/test_polars.py

```diff
@@ -262,18 +253,114 @@ def test_onceframe_multi_collect():
 
 
 def test_onceframe_lazy():
-    dp.enable_features("rust-stack-trace")
     pl = pytest.importorskip("polars")
 
-    lf_domain, lf = example_lf(
-        margin=["B"], public_info="keys", max_partition_length=50
-    )
-    
-    expr = pl.col("A").fill_null(1.5).dp.sum((1.0, 2.0), scale=0.0)
-    plan = seed(lf.schema).group_by("B").agg(expr).sort("B")
+    lf_domain, lf = example_lf()
+    plan = seed(lf.schema).select(pl.len().dp.noise(0.0))
     m_lf = dp.m.make_private_lazyframe(
         lf_domain, dp.symmetric_distance(), dp.max_divergence(T=float), plan
     )
 
     of = m_lf(lf)
+    dp.enable_features("honest-but-curious")
     assert isinstance(of.lazy(), pl.LazyFrame)
+
+
+@pytest.mark.parametrize(
+    "measure", [dp.max_divergence(T=float), dp.zero_concentrated_divergence(T=float)]
+)
+def test_mechanisms(measure):
+    pl_testing = pytest.importorskip("polars.testing")
+
+    pl = pytest.importorskip("polars")
+
+    lf_domain, lf = example_lf()
+
+    if measure == dp.max_divergence(T=float):
+        expr = pl.len().dp.laplace(0.0)
+    else:
+        expr = pl.len().dp.gaussian(0.0)
+
+    plan = seed(lf.schema).select(expr)
+    m_lf = dp.m.make_private_lazyframe(
+        lf_domain, dp.symmetric_distance(), measure, plan, 0.0
+    )
+
+    expect = pl.DataFrame([pl.Series("len", [50], dtype=pl.UInt32)])
+    pl_testing.assert_frame_equal(m_lf(lf).collect(), expect)
+
+
+def test_polars_context():
+    pl = pytest.importorskip("polars")
+
+    lf = pl.LazyFrame(
+        {"A": [1, 2, 3, 4], "B": ["x", "x", "y", None]},
+        schema={"A": pl.Int32, "B": pl.String},
+    )
+
+    context = dp.Context.compositor(
+        data=lf,
+        privacy_unit=dp.unit_of(contributions=1),
+        privacy_loss=dp.loss_of(epsilon=1.0),
+        split_evenly_over=2,
+        margins={
+            # TODO: this is redundant with the second margin
+            (): dp.Margin(max_partition_length=5),
+            ("B",): dp.Margin(public_info="keys", max_partition_length=5),
+        },
+    )
+
+    (
+        context.query()
+        .with_columns(pl.col("B").is_null().alias("B_nulls"))
+        .filter(pl.col("B_nulls"))
+        .select(pl.col("A").fill_null(2.0).dp.sum((0, 3)))
+        .release()
+        .collect()
+    )
+
+    (
+        context.query()
+        .group_by("B")
+        .agg(pl.len().dp.noise(), pl.col("A").fill_null(2).dp.sum((0, 3)))
+        .release()  # type: ignore[union-attr]
```

Comment:
It isn't ideal, and is unfortunately not an easy fix. The reason why the typing generally works out so nicely is because most of the LazyFrame APIs return `Self`, which, when subclassed, is `LazyFrameQuery`. However `.group_by` returns `LazyGroupBy`, which exits our subclass.

Reply:
- Filed #1685